### PR TITLE
[1/1] Support resource-class flag in eas-cli

### PR DESCRIPTION
**Pull request by christopherwalter**
# Why
We want to add a flag to allow users to specify resource-class when they
run `eas build`. The flag is experimental for now, and we haven't
decided exactly how we'll incorporate the resource classes into plan
pricing. But this PR doesn't deal with that - it just makes the
necessary changes to eas-cli to support specifying a resource class when
running `eas build`.

# How
 Add a new flag `resource-class` whose options are 'large | default', 
and translate the user-provided value into platform-specific BuildResourceClass 
values for each build job we create. Then we send each job to WWW GraphQL
mutations with their BuildResourceClass included.

# Test Plan

See this comment for the e2e test plan: https://github.com/expo/universe/pull/9800#issuecomment-1163988895

### CODE REVIEW

**Comment by christopherwalter** [comment]
/changelog-entry [new-feature] [Add `--build-resource-class` flag for EAS Build commands.]

File: packages/eas-cli/graphql-codegen.yml
Comment (wkozyra95) [inline packages/eas-cli/graphql-codegen.yml:2]: remember to revert before merging

File: packages/eas-cli/src/commands/build/index.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:95]: This is not really consistent with how we name stuff, I would suggest naming it lower case without underscores and platform
"default" | "large" | "m1"

File: packages/eas-cli/src/commands/build/index.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:142]: m1 is obviously only for ios, but there is no reason to not support default/large for all option

File: packages/eas-cli/src/commands/build/index.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:131]: You don't need Map here, it can be just a plain object

File: packages/eas-cli/src/build/types.ts
Comment (wkozyra95) [inline packages/eas-cli/src/build/types.ts:16]: you are using this type in some places and in others, you are importing a copy of it from eas-build-job, it should be more consistent in either direction

File: packages/eas-cli/src/commands/build/index.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:93]:
```suggestion
    'resource-class': Flags.enum({
```

**Review by dsokal — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/commands/build/index.ts
Comment (christopherwalter) [reply packages/eas-cli/src/commands/build/index.ts:142]: It seems like a nice-to-have to me, at least for right now. The same functionality is achieved by just running two `eas build` commands, one for android and one for iOS. I'm curious if @brentvatne has other thoughts on this.

File: packages/eas-cli/src/commands/build/index.ts
Comment (christopherwalter) [reply packages/eas-cli/src/commands/build/index.ts:131]: I'm going to remove this logic anyway, in light of making the flag options non-platform specific. I'll move the resource-class determination into the existing buildProfile generation logic, which should make things a bit cleaner.

File: packages/eas-cli/src/commands/build/index.ts
Comment (christopherwalter) [reply packages/eas-cli/src/commands/build/index.ts]: fixed

**Comment by christopherwalter** [comment]
Right now this test is failing locally but I'm not quite sure why: `packages/eas-cli/src/commandUtils/__tests__/EasCommand-test.ts`

Here's the failure message:

```
● EasCommand › without exceptions › ensures the user is logged in

  TypeError: The "warning" argument must be of type string or an instance of Error. Received an instance of Error

    at Config.warn (node_modules/@oclif/core/lib/config/config.js:470:21)
    at node_modules/@oclif/core/lib/config/config.js:448:22
        at async Promise.all (index 1)
    at Config.loadPlugins (node_modules/@oclif/core/lib/config/config.js:423:9)
    at Config.loadCorePlugins (node_modules/@oclif/core/lib/config/config.js:150:13)
    at Config.load (node_modules/@oclif/core/lib/config/config.js:141:9)
    at Function.load (node_modules/@oclif/core/lib/config/config.js:80:9)
    at Function.Object.<anonymous>.Command.run (node_modules/@oclif/core/lib/command.js:183:20)
    ```
    
   I haven't really dug into it yet, but let me know if you have any ideas @dsokal @wkozyra95.

File: packages/eas-cli/src/commands/build/index.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:153]:
```suggestion
    if (flags['resource-class'] && !Object.values(UserInputResourceClass).includes(flags['resource-class'])) {
      Errors.error(`Invalid resource-class: '${flags['resource-class']}'`, { exit: 1 });
    }
```

The code you wrote is really tricky and hard to understand. Also, if you wanted to add a new value to the enum you would need to know about this code.

File: packages/eas-json/src/errors.ts
Comment (dsokal) [inline packages/eas-json/src/errors.ts:9]: Why is this error class in `eas-json` package? If we decided to go for the CLI param, then it has nothing to do with eas.json.

File: packages/eas-cli/src/utils/profiles.ts
Comment (dsokal) [inline packages/eas-cli/src/utils/profiles.ts:18]: Profiles are an abstraction for entities from eas.json. This code does not belong here.

File: packages/eas-cli/src/utils/profiles.ts
Comment (dsokal) [inline packages/eas-cli/src/utils/profiles.ts:44]: This is not really readable. It's better to have a map like:

```
const xyz: Record<Platform, Record<UserInputResourceClass, BuildResourceClass>> = {
  [Platform.ANDROID]: {
    [UserInputResourceClass.DEFAULT]: BuildResourceClass.AndroidDefault,
    // ..
  },
  // ..
};
```

And then try to access the appropriate value from map:
```
const aaa = xyz[platform]?.[userInputResourceClass];
if (!aaa) {
  throw new Error('blah blah');
}
```

My approach is better because:
- (!) If you ever add a new value to `Platform` or `UserInputResourceClass`, TS will complain that some mappings are missing in `xyz` map.
- The code is more straightforward.

File: packages/eas-cli/src/commands/build/index.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:87]:
```suggestion
      description: 'The resource class that will be used to run this build [experimental]',
```

File: packages/eas-cli/src/build/createContext.ts
Comment (dsokal) [inline packages/eas-cli/src/build/createContext.ts:38]: This shouldn't be optional here. We already know the exact value after converting user input to graphql enum.

File: packages/eas-cli/src/build/context.ts
Comment (dsokal) [inline packages/eas-cli/src/build/context.ts:34]: This shouldn't be optional here. We already know the exact value after converting user input to graphql enum.

File: packages/eas-cli/src/build/build.ts
Comment (dsokal) [inline packages/eas-cli/src/build/build.ts:75]: Resource class should be always defined in this context.

**Review by dsokal — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/build/runBuildAndSubmit.ts
Comment (christopherwalter) [reply packages/eas-cli/src/build/runBuildAndSubmit.ts:70]: Yeah I felt like that was sketchy to associate it with profile. Good call.

File: packages/eas-json/src/errors.ts
Comment (christopherwalter) [reply packages/eas-json/src/errors.ts:9]: I have no good answer for this 😅 I just grep'd for *Error and figured this is where error definitions go...

I'll remove this since we're not using it anymore.

File: packages/eas-cli/src/utils/profiles.ts
Comment (dsokal) [inline packages/eas-cli/src/utils/profiles.ts:41]: Revert

**Review by dsokal — APPROVED** [review]
LGTM 🎉 

- Add a changelog entry.
- Before we merge let's wait for a review from @wkozyra95

File: packages/eas-cli/src/commands/build/index.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:146]: do you need to validate that? I thought that ocliff is handling that out of the box?

File: packages/eas-cli/src/commands/build/index.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:87]: resources class term is common in CI ecosystem, but for users it might not be obvious what it means. It's fine as a name of the option, but the description should use sth else to define it e.g `The instance type that will be used to run this build`

**Review by wkozyra95 — APPROVED** [review]
I do not see any validation before the build starts that checks if the user is allowed to run on instance x. I think that throwing an error when build request is send is to late for that, user should get feedback on that at the very beginning of the process

File: packages/eas-cli/src/commands/build/index.ts
Comment (christopherwalter) [reply packages/eas-cli/src/commands/build/index.ts:146]: ah I didn't realize that was the case, I'll remove this

**Comment by christopherwalter** [comment]
> I do not see any validation before the build starts that checks if the user is allowed to run on instance x. I think that throwing an error when build request is send is to late for that, user should get feedback on that at the very beginning of the process

@wkozyra95 unfortunately we can only do this check on the server-side. I could create a new endpoint and send an earlier request to validate this, but I think that's out-of-scope for now.

**Comment by christopherwalter** [comment]
/changelog-entry new-feature Add experimental `--resource-class` flag for Build commands

**Pull request merged**
